### DISCUSSION TITLE
Should we be using sizeof(buf) or sizeof (buf)

### DISCUSSION THREAD

**Discussion by hughsie** [fwupd/fwupd#3541] [General]
@superm1 @djcampello 

At the moment in the codebase we use both `sizeof(buf)` and `sizeof (buf)` -- in my head the first makes the most sense as a it's a preprocessor directive, but on the other hand it just looks like a function to most people.

What do we think about doing a bulk conversion of the mixed format we have now to either the former or the latter. I know it's baby steps, but I thought we could start with some of the lowest hanging fruit.

**Comment by hughsie** [comment]
Of course, maybe the argument might be to go full LKML style, i.e. with `no(spaces(between(the(brackets)))))` -- this certainly helps for the uncommon case of the odd-looking `fu_whatever_get_array (self)->len`

**Comment by hughsie** [comment]
Interesting GLib has the same issue:

[hughsie@fedora glib (master %)]$ git grep "sizeof (" | wc -l
1419
[hughsie@fedora glib (master %)]$ git grep "sizeof(" | wc -l
319

**Comment by hughsie** [comment]
It turns out not many people agree with my space before bracket, except for sizeof rule :) https://twitter.com/hughsient/status/1418957839503134725

**Reply by superm1** [reply]
I think that's my preference too.

**Reply by djcampello** [reply]
Same for me :)

**Reply by hughsie** [reply]
It feels like changing my card signature, but fwupd isn't just my pet project anymore: https://github.com/fwupd/fwupd/pull/3542

Do either of you know how to wire that into precommit just for the changed lines? e.g. I still think we should port incrementally rather than breaking the out-of-tree patches for everybody.

**Reply by djcampello** [reply]
Without the need to treat different sizeof/gautoptr/gfree spacing we do not need the contrib/reformat-code.py... `git clang-format` should do what you want. Not sure how to wire it to precommit though.

```
$ git clang-format -h
usage: git clang-format [OPTIONS] [<commit>] [<commit>] [--] [<file>...]

If zero or one commits are given, run clang-format on all lines that differ
between the working directory and <commit>, which defaults to HEAD.  Changes are
only applied to the working directory.
...
```

**Reply by djcampello** [reply]
I imagine should be similar to: 

```
-   repo: https://github.com/ambv/black
    rev: 21.6b0
    hooks:
    - id: black
```

in https://github.com/fwupd/fwupd/commits/master/.pre-commit-config.yaml

**Reply by hughsie** [reply]
Using:

	-   repo: git://github.com/doublify/pre-commit-clang-format
	    rev: 62302476d0da01515660132d76902359bed0f782
	    hooks:
	    - id: clang-format

it seems to reformat all of any file it touches, not just the changed hunks.

**Reply by djcampello** [reply]
What about:

```
    - id: clang-format
      name: clang-format
      language: script
      entry: ./contrib/refomat-code.py
```

If the script is executed without parameters it will try to format the diff with HEAD. Not sure if this works on pre-commit as there is already a commit?

**Reply by superm1** [reply]
> What about:
> 
> ```
>     - id: clang-format
>       name: clang-format
>       language: script
>       entry: ./contrib/refomat-code.py
> ```
> 
> If the script is executed without parameters it will try to format the diff with HEAD. Not sure if this works on pre-commit as there is already a commit

My thought is:
1. Add to precommit but comment it out for CI
2. a "separate" CI job that checks the style of the diff relative to master.

**Reply by djcampello** [reply]
If we add the script to `.pre-commit-config.yaml` it is called with the list of all the files that changed which makes it behave as a whole file formatter. The existent script needs to be modified to be used as a pre-commit hook by getting rid of the whole-file formatting (as long as the fixups for gsize etc)...

**Reply by djcampello** [reply]
Here is what I meant: https://github.com/fwupd/fwupd/commit/972fd77ac90dfc5497cfda19b75d1fff697340f3

**Reply by superm1** [reply]
> 972fd77

I cherry-picked that on top of https://github.com/fwupd/fwupd/pull/3542 and yes, experimenting it seems to do the right thing, at least for a single commit.  I think it will work for local purpose, but not for CI.  I took what you did and put it on 3542, and then added what I think is the remaining fix to make CI work, take a look.

**Comment by YehezkelShB** [comment]
What about the option of no brackets at all for `sizeof`? :)
(Ah, it doesn't work for types, only for expressions, so you still have to decide what to do when applying it to a type)
